File: 543/DiameterOfBinaryTree.scala
object DiameterOfBinaryTree {
  class TreeNode(val _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }

  def diameterOfBinaryTree(root: TreeNode): Int = {
    var res = 0

    def loop(root: TreeNode): Int = if (root == null) 0 else {
      val l = loop(root.left)
      val r = loop(root.right)
      res = res max (l + r)
      (l max r) + 1
    }

    loop(root)
    res
  }

}

File: 543/diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
    let res = 0;

    function dfs(root) {
        if (!root) return 0;
        let l = dfs(root.left);
        let r = dfs(root.right);
        res = Math.max(res, l + r);
        return Math.max(l, r) + 1;
    }

    dfs(root);
    return res;
}